Modules/Loadable/Markups/Widgets/qMRMLMarkupsToolBar.cxx

```diff
@@ -0,0 +1,530 @@
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+  This file was originally developed by Julien Finet, Kitware Inc.
+  and was partially funded by NIH grant 3P41RR013218-12S1
+
+==============================================================================*/
+
+// Qt includes
+#include <QDebug>
+#include <QToolButton>
+#include <QMenu>
+#include <QCheckBox>
+#include <QSignalMapper>
+#include <QSplitter>
+#include <QShortcut>
+#include <QKeySequence>
+#include <QTest>
+
+// Slicer includes
+#include "qSlicerCoreApplication.h"
+#include "qSlicerApplication.h"
+#include "qSlicerLayoutManager.h"
+
+// MRML includes
+//#include "qMRMLMarkupsToolBar.h"
+#include "qMRMLMarkupsToolBar_p.h"
+#include "qMRMLNodeComboBox.h"
+#include "qMRMLThreeDView.h"
+#include "qMRMLThreeDWidget.h"
+#include "qMRMLSliceView.h"
+#include "qMRMLSliceWidget.h"
+#include <vtkMRMLScene.h>
+#include <vtkMRMLMarkupsNode.h>
+#include <vtkMRMLInteractionNode.h>
+#include <vtkMRMLSelectionNode.h>
+#include <vtkMRMLLayoutLogic.h>
+#include <vtkMRMLSelectionNode.h>
+#include <vtkMRMLSliceNode.h>
+#include <vtkMRMLViewNode.h>
+#include <vtkMRMLWindowLevelWidget.h>
+#include <qSlicerMarkupsPlaceWidget.h>
+
+// SlicerLogic includes
+#include <vtkSlicerApplicationLogic.h>
+// MRMLLogic includes
+#include <vtkMRMLApplicationLogic.h>
+#include <vtkSlicerMarkupsLogic.h>
+
+// VTK includes
+#include <vtkWeakPointer.h>
+#include <vtkSmartPointer.h>
+
+//---------------------------------------------------------------------------
+qMRMLMarkupsToolBarPrivate::qMRMLMarkupsToolBarPrivate(qMRMLMarkupsToolBar& object)
+  : q_ptr(&object)
+{
+  this->MarkupsNodeSelector = 0;
+  this->MarkupsPlaceWidget = 0;
+  //this->PlaceModesActionGroup = nullptr;
+  this->DefaultPlaceClassName = "vtkMRMLMarkupsFiducialNode";
+
+}
+
+//---------------------------------------------------------------------------
+void qMRMLMarkupsToolBarPrivate::init()
+{
+  Q_Q(qMRMLMarkupsToolBar);
+
+  // Markups node selector
+  this->MarkupsNodeSelector = new qMRMLNodeComboBox();
+  this->MarkupsNodeSelector->setNodeTypes(QStringList(QString("vtkMRMLMarkupsNode")));
+  this->MarkupsNodeSelector->setNoneEnabled(false);
+  this->MarkupsNodeSelector->setAddEnabled(false);
+  this->MarkupsNodeSelector->setRenameEnabled(true);
+  this->MarkupsNodeSelector->setMaximumWidth(150);
+  this->MarkupsNodeSelector->setEnabled(true);
+
+  q->addWidget(this->MarkupsNodeSelector);
+
+  connect(this->MarkupsNodeSelector, SIGNAL(currentNodeChanged(vtkMRMLNode*)),
+    q, SIGNAL(activeMarkupsNodeChanged(vtkMRMLNode*)) );
+  connect(this->MarkupsNodeSelector, SIGNAL(currentNodeChanged(vtkMRMLNode*)), q, SLOT(onMarkupsNodeChanged()));
+
+  // set up keyboard shortcuts
+  q->addCreateNodeShortcut("Shift+A");
+  q->addTogglePersistenceShortcut("Shift+Z");
+  q->addPlacePointShortcut("Space");
```

Comment:
If I'm typing to rename a node and I rename the node "A Zoo" will it create a node, toggle persistence and toggle placement mode?

Reply:
Thanks @jamesobutler, these shortcuts work regardless of the mouse placement, but if a text field is active then the shortcuts are blocked, so this should not be an issue. Three button shortcuts that we tested were not easy to use from a workflow standpoint, but I could replace Shift with Ctrl to make it more distinct.

Reply:
I would prefer not to use single keys or Shift+keys or space as keyboard shortcuts in Slicer core, because it could cause conflicts with more specialized workflows and modules, which want to define their own, simplest-possible shortcuts. For example, space is already used in Segment Editor for toggling between last two segment editor effects - using the same key for markups placement enable/disable would probably not work at all. These kind of shortcuts can be temporarily added in a dedicated landmarking module (similarly how Segment Editor module temporarily adds its shortcuts).